johnzon-core/src/main/java/org/apache/johnzon/core/Types.java

```diff
@@ -0,0 +1,125 @@
+        }
+        if (type instanceof ParameterizedType) {
+            return resolveArgumentTypes((ParameterizedType) type, parameterizedClass);
+        }
+        throw new IllegalArgumentException("Cannot resolve argument types from " + type.getClass().getSimpleName());
+    }
+
+    private static Type[] resolveArgumentTypes(Class<?> type, Class<?> parameterizedClass) {
+        if (parameterizedClass.equals(type)) {
+            // May return Class[] instead of Type[], so copy it as a Type[] to avoid
+            // problems in visit(ParameterizedType)
+            return Arrays.copyOf(type.getTypeParameters(), parameterizedClass.getTypeParameters().length, Type[].class);
+        }
+        if (type.getSuperclass() != null && parameterizedClass.isAssignableFrom(type.getSuperclass())) {
+            return resolveArgumentTypes(type.getGenericSuperclass(), parameterizedClass);
+        }
+        Class<?>[] interfaces = type.getInterfaces();
+        Type[] genericInterfaces = type.getGenericInterfaces();
+        for (int i = 0; i < interfaces.length; i++) {
+            if (parameterizedClass.isAssignableFrom(interfaces[i])) {
+                return resolveArgumentTypes(genericInterfaces[i], parameterizedClass);
+            }
+        }
+        throw new IllegalArgumentException(String.format("%s is not assignable from %s", type, parameterizedClass));
+    }
+
+    private static Type[] resolveArgumentTypes(ParameterizedType type, Class<?> parameterizedClass) {
+        Class<?> rawType = (Class<?>) type.getRawType(); // always a Class
+        TypeVariable<?>[] typeVariables = rawType.getTypeParameters();
+        Type[] types = resolveArgumentTypes(rawType, parameterizedClass);
+        for (int i = 0; i < types.length; i++) {
+            if (types[i] instanceof TypeVariable<?>) {
+                TypeVariable<?> typeVariable = (TypeVariable<?>) types[i];
+                for (int j = 0; j < typeVariables.length; j++) {
+                    if (typeVariables[j].getName().equals(typeVariable.getName())) {
+                        types[i] = type.getActualTypeArguments()[j];
+                    }
+                }
+            }
+        }
+        return types;
+    }
+
+    private Types() {
+        // no-op
+    }
+    
+    private static class ParameterizedTypeImpl implements ParameterizedType {
+
+        private final Type rawType;
+        private final Type[] arguments;
+
+        ParameterizedTypeImpl(Type rawType, Type... arguments) {
+            this.rawType = rawType;
+            this.arguments = arguments;
+        }
+
+        @Override
+        public Type getRawType() {
+            return rawType;
+        }
+
+        @Override
+        public Type getOwnerType() {
+            return null;
+        }
+
+        @Override
+        public Type[] getActualTypeArguments() {
+            return arguments;
+        }
+
+        // TODO equals, hashcode, toString if needed

```

Comment:
side note: not mandatory today but if desired you can steal these methods from https://github.com/apache/johnzon/blob/master/johnzon-mapper/src/main/java/org/apache/johnzon/mapper/reflection/JohnzonParameterizedType.java (don't link the two modules since mapper should run with another json-p impl IIRC)